Add "Text & Image" Block

jorgefilipecosta opened the pull request:
# Description

Context: #6993

This PR implements a block that contains two areas one for a media element (image/video) and one area for blocks (buttons, paragraphs etc). The elements of both areas are vertically aligned.
This PR contains an alternative to https://github.com/WordPress/gutenberg/pull/7414. So we can compare the two approaches side by side.
Here instead of middle blocks, the parent block will self-contain the media area (that allows images or videos), so we have only one simple nested area for the content.

## Screenshots
![sep-04-2018 18-49-22](https://user-images.githubusercontent.com/11271197/45048364-a02f9280-b073-11e8-8434-60cc989af4bc.gif)


![image](https://user-images.githubusercontent.com/11271197/45048283-678fb900-b073-11e8-965e-cd9c7abf9704.png)



## Reviewer notes:
Files under packages/editor/src/utils/media-upload and packages/editor/src/hooks/align.js should not be reviewed in this PR as they are just cherry picks from other PRs: https://github.com/WordPress/gutenberg/pull/9707, https://github.com/WordPress/gutenberg/pull/9704, and https://github.com/WordPress/gutenberg/pull/9634.


Depends on: https://github.com/WordPress/gutenberg/pull/9707, https://github.com/WordPress/gutenberg/pull/9704, and https://github.com/WordPress/gutenberg/pull/9634

jorgefilipecosta committed:
Implement Media & Text Block

ZebulanStanphill commented on packages/block-library/src/layout-half-media-text/index.js:
```diff
@@ -0,0 +1,122 @@
+	},
+
+	supports: {
+		align: [ 'center', 'wide', 'full' ],
```
Why is center alignment supported? What would that even do for this block?

jorgefilipecosta replied:
By default, we have a wide alignment for this block the block goes out of the content area if the user does not want that, the center alignment (the default on most other blocks) can be used.

ZebulanStanphill replied:
@jorgefilipecosta Center is not the default alignment for most blocks. No alignment is the default alignment. There is a difference. The Columns block supports wide and full alignments but not the center alignment because the center alignment does not do anything for it.

jorgefilipecosta replied:
The approach I was trying to follow of using center align was a way to unset the alignment is confusing I followed your suggestion and used just wide and full alignments for things to work correctly I needed to apply some engine fixes: https://github.com/WordPress/gutenberg/pull/9634

ZebulanStanphill commented on packages/block-library/src/layout-half-media-text/edit.js:
```diff
@@ -0,0 +1,145 @@
+const MEDIA_POSITIONS = [ 'left', 'right' ];
+const ALLOWED_BLOCKS = [ 'core/button', 'core/paragraph', 'core/heading', 'core/list' ];
+const TEMPLATE = [
+	[ 'core/paragraph', { fontSize: 'large', placeholder: 'Content...' } ],
```
This should use an ellipsis (…) rather than 3 periods.

jasmussen commented:
Love the most recent changes. This is going to be a great block. Love:

- That the block is simple — child blocks that aren't manipulable aren't surfaced. 
- Easy to add images, easy to resize, easy to add text, easy to configure the block

It's delicious!

The drag handle to resize needs a little love. Can we make sure the drag handle is only visible when the block is selected first?

In addition to this, can we change it to use the same resizing handle that images do? I.e. instead of this:

<img width="1200" alt="screen shot 2018-09-05 at 09 05 41" src="https://user-images.githubusercontent.com/1204802/45076756-8af64a80-b0eb-11e8-8b36-d83f201051e8.png">

We do this:

![this](https://user-images.githubusercontent.com/1204802/45076838-c5f87e00-b0eb-11e8-82df-b00790530486.png)

And instead of this:

<img width="1256" alt="screen shot 2018-09-05 at 09 08 33" src="https://user-images.githubusercontent.com/1204802/45076851-cb55c880-b0eb-11e8-8890-2bb79a672f7e.png">

We do this:

![this2](https://user-images.githubusercontent.com/1204802/45076885-e4f71000-b0eb-11e8-85cd-f5c48dee6ba4.png)

I'd also personally be okay if the placeholder spot wasn't resizable, and it wasn't until a media item was added that the resize handle appeared. But it's fine either way.

jorgefilipecosta commented:
Thank you for your reviews @ZebulanStanphill and @jasmussen all the concerns were addressed :)

jasmussen commented:
This is really nice, I'm digging this block a lot:

<img width="1284" alt="screen shot 2018-09-06 at 10 46 28" src="https://user-images.githubusercontent.com/1204802/45145945-23f79500-b1c2-11e8-90de-f187a7c4237c.png">

However, possibly a rebase issue, but I'm seeing the title spot way smaller font size than in master:

<img width="514" alt="screen shot 2018-09-06 at 10 45 51" src="https://user-images.githubusercontent.com/1204802/45145981-32de4780-b1c2-11e8-99e4-bd36c641f319.png">

Actually that's also in master.. I will investigate that separately. But 👍 👍 on the design! All this needs is a code review update and good to go I'd say.

ZebulanStanphill commented:
Nice! Everything looks good, except for two things.

First of all, is the image supposed to extend outside of the container once you increase it to a really large size?
![image](https://user-images.githubusercontent.com/19592990/45307767-7ea43000-b4e5-11e8-8ef0-5760dbe2b70d.png)

Second, the blocks does not have any default responsive styling. Of course, themes could provide styling, but I wonder if Gutenberg should provide default styles that themes could opt-out from.

I know right now Gutenberg has opt-in styles for blocks, but that seems very backwards to me. Gutenberg should not be providing any opt-in styles... only opt-out styles which exist for compatibility with older themes. Newer themes should be expected to provide most, if not all the styling for blocks. Perhaps even the "essential" styles for blocks should be opt-out so that themes can get maximum control over how blocks appear.

mtias commented:
@jorgefilipecosta in Spotlight mode we are losing the ability to resize the image.

mtias commented:
> I know right now Gutenberg has opt-in styles for blocks, but that seems very backwards to me. 

This is not entirely correct. Structural styles are loaded for all themes. It's visual styles — which have been separated through the `theme.scss` files — the ones that are opt-in. They include things like the quote left border, typographic opinions and certain colors, etc. And it's necessary for it to be this way because otherwise enabling Gutenberg could visually break a theme's expectations.

New themes that want to work with Gutenberg out of the box would accept these styles and build on top of them.

jorgefilipecosta commented:
> @jorgefilipecosta in Spotlight mode we are losing the ability to resize the image.

Hi @mtias, nice catch this problem has solved.

ZebulanStanphill commented:
@mtias Having `theme.scss` files at all is a strange decision. There should not be any `theme.scss` because new themes should be providing their own styles for blocks anyway, should they not? Gutenberg should not take on the responsibility of non-essential styling. The opt-in styles should be in a new default theme or something, shouldn't they? Putting it in core feels like bloat to me.

jasmussen commented:
Appreciate your passion, @ZebulanStanphill, it's visceral 💕 

In this case, discussing the theme.scss file is sort of rehashing an old discussion, and it would best to keep the discussion on the older tickets. Not only so we don't have to explain it again — that's not scalable — but to keep the context and history together. The discussion starts in #2905 (old ticket, as you can see), was refined in #5360, and there are some future ideas in #5445. 

To summarize briefly, the most basic example I can give is of the Separator block.

Most themes style the `hr` tag. Gutenberg also styles this, by making it 2px tall, centered, and not full-width. Because most themes already have a style for this, the CSS that makes the separator 2px tall, centered, and not full width is stored in `theme.scss`. That way themes need to explicitly say "yes, give me that style".

But the Separator block also has two variations, one which is full wide and thin, and one which is just 3 dots. No WordPress theme has styled these yet, because they are new. That's why those are stored in `style.scss` and are given to themes automatically. That way if a user uses TwentySixteen which has no concept of blocks, those variation styles will still work fine, without the theme having to be updated.

Ultimately it's about allowing Gutenberg to provide some well-designed blocks out of the box, ensuring that expectations are met between what a user sees in the editor and sees on the front-end, and giving theme developers options at the same time. This way, a freshly built theme that has an editor style and is Gutenberg aware will have lots of options of ensuring consistency between the frontend and the backend, providing their own styles where they want to. But at the same time every existing WordPress theme which has no concept of blocks can have some measure of consistency between the frontend and backend also, without having to do anything.

jasmussen commented:
I pushed a little tweak to increase the space between image and text. It looks nicer.

But I also noted a quick bug that somehow I missed previously. This block isn't responsive:

![bug](https://user-images.githubusercontent.com/1204802/45409546-c0f97880-b66f-11e8-8d0c-180f9c662575.gif)

It's due to the explicit _width_ set on the image. I'm trying to figure out how we can make it responsive, and it's tricky. On one hand it's about meeting user expectations — if it's resizable, then for sure reflect what I see. But at the same time, we have to make this responsive.

Perhaps using percent can work? I'm going to try and see if I can get that to work.

jasmussen commented:
Okay so the responsiveness here is going to be tricky. If we remove the resizing feature entirely, it's trivial:

![responsive](https://user-images.githubusercontent.com/1204802/45410029-0ff3dd80-b671-11e8-9717-c07fcfd8f1c1.gif)

This is because with the grid layout, `grid-template-columns: auto fit-content(50%);` takes care of everything.

Which means if we _are_ to have a resizing feature, which I think is killer, we need the resize handle to change that rule instead. Is that even possible?

jorgefilipecosta commented:
> This is because with the grid layout, grid-template-columns: auto fit-content(50%); takes care of everything.

One simple option to get a responsive layout would be to use a media query. That makes the layout appear in two rows (using a different value for grid-template-areas) if the size is bellow a certain threshold. Themes can also easily do that, making sure this block is responsive when using this layout.
I can add a commit with this so we can test how it looks.

> Which means if we are to have a resizing feature, which I think is killer, we need the resize handle to change that rule instead. Is that even possible?

I guess we can change the rule by applying an inline style:
e.g: 'grid-template-columns: auto WIDTHpx'

jasmussen commented:
> I guess we can change the rule by applying an inline style:
> e.g: 'grid-template-columns: auto WIDTHpx'

I like this option the most. But can we use % instead of px here?

jorgefilipecosta commented:
> I like this option the most. But can we use % instead of px here?

Unfortunately, % here seems very tricky. When using a % the resizer component computes the % of space in relation to the parent of the object we are resizing. But in this case, the parent of the object we are resizing is the grid cell whose width depends on the child. In my tries to use %, I got the idea our resizer component can deal with configuring size in % but when we are not using a flex or grid layout that already manages the dimensions.
Just to confirm, this responsive system would never put the media and content on different lines. They would always be side by side, but instead of the media area having a fixed size, its size would be a fixed % of the size available for the media block? Is that right?

jasmussen commented:
> Unfortunately, % here seems very tricky. When using a % the resizer component computes the % of space in relation to the parent of the object we are resizing. But in this case, the parent of the object we are resizing is the grid cell whose width depends on the child. 

Yes, I noticed that. Which means we need to resize the grid-cell itself, right?

> Just to confirm, this responsive system would never put the media and content on different lines. They would always be side by side, but instead of the media area having a fixed size, its size would be a fixed % of the size available for the media block? Is that right?

Correct.

I think we _could_ explore additional responsive measures like stacking on top on mobile, I think that would be fine, but can totally be done separately. Right now the real issue here is that even on desktop breakpoints, it's not responsive due to the inline pixel width of any media item resized using the size handler. 

So say you are on a big 27 inch display with a fullscreen browser, and you have this lovely half media block with text on the right and it's looking great. Now you resize the browser, but because the image was sized according to what you saw on the big  when you looked at it on the big screen, suddenly it's way too big and the text becomes a thin sliver on the right.

jasmussen commented:
THAT IS AMAZING HOLY GUACAMOLE wow so well done. You get a medal: 🏅

I can't test right now, can test later. But if it works as well as it looks like, ship it!

jasmussen reviewed (approved):
Just tested. I can't believe you did it. This is completely amazing.

A warm and grateful 👍 👍 to this PR from the design and experience side.

Still probably needs a final code review and a rebase. But otherwise cool.

One final thing, doesn't have to be done — but I noticed that as soon as you insert this block using the slash command, focus is not set on the first text block, the one with large text, but on the second one. Is this what we want? Especially now that the editor itself sets focus on the title field on load.

jasmussen commented:
Not a blocker for this PR, I'm excited to get this in.

But it seems like the consecutive ⌘A behavior isn't working here if this block is the only block on the page. Though perhaps that's an issue with other media blocks as well?

aduth commented:
Forgive me if I've missed the context (I tried skimming / searching), but... is there any reason this can't be just a Columns block?

- Is it a workflow / convenience thing around the setup required to get a Columns block to match the appearance offered by this? If so, couldn't we tackle this from the angle of having some concept of block scaffolding, where it's still saved as the Columns block, but with additional user conveniences, even if that means it _appears_ as its own block in the block inserter.
- Is it a technical limitation of the Columns block in e.g. supporting column resizing / responsiveness. If so, aren't these enhancements we should consider for the Columns block (and would probably want anyways)?

I might imagine a counterpoint being that with `InnerBlocks` managing most of the complexity already anyways, these sorts of layout blocks are likely to be thin layers. I just wonder if the points above might be things we want anyways, irrespective of what we're building here.

jasmussen commented:
> Forgive me if I've missed the context (I tried skimming / searching), but... is there any reason this can't be just a Columns block?

> Is it a workflow / convenience thing around the setup required to get a Columns block to match the appearance offered by this? If so, couldn't we tackle this from the angle of having some concept of block scaffolding, where it's still saved as the Columns block, but with additional user conveniences, even if that means it appears as its own block in the block inserter

> Is it a technical limitation of the Columns block in e.g. supporting column resizing / responsiveness. If so, aren't these enhancements we should consider for the Columns block (and would probably want anyways)?

Responding to all those at once because they are intricately related. I know @mtias has additional fleshed out thoughts on the role these play in a more distant future, but he's AFK so I'll do my best in his absence.

Also to the 2nd quote, I won't speak too much to the technical structure, just to the user experience. Whether this is technically a columns block or something else under the hood, I trust you all to use the best tools and approaches available. 

The important thing here is that it is **simple**. 

The thing is, putting a video on the left and text on the right (vertically centered) is a seemingly trivial layout, but it is surprisingly complicated to achieve, even for someone with access to custom HTML and CSS. 

Sure, you could probably put it together using Gutenblocks already — a 2 column block with an embed or image on the left and text and other on the right — but it is not trivial, and we _want_ it to be trivial.

Many of our Gutenblocks are akin to LEGO blocks. You can put several of them together and create a wonderful layout. But every once in a while, there's a thing you want to build in LEGO that isn't that easy to build. 

Forgive me in advance for this metaphor, but I'm Danish, LEGO is Danish, and I'm finding myself unable to stop this metaphor. Here we go, say you need a LEGO tree. Sure, you could put together a stack of green _ones_ with some flatter blocks up top to form a tree-crown:

![screen shot 2018-09-19 at 17 57 19](https://user-images.githubusercontent.com/1204802/45765569-98f0b300-bc35-11e8-987f-fc42f6887ca0.png)

But in this case it's quite nice that a LEGO _tree_ exists:

![tree](https://user-images.githubusercontent.com/1204802/45765599-a9089280-bc35-11e8-9f86-c4b2b313ef9a.png)

It is fair and correct to note that the green LEGO block is more universal and there for _way more widely applicable_, but for those odd cases, some of these highly detailed and highly specialized LEGO are exactly what you need. 

In this case, the fact that we're making a dedicated "Media & Text" block (by the way should we call it that instead of "Half Media"?) makes it trivial to achieve nice looking layouts in any situation, for anyone. You could even use this block inside a columns block, or in other contexts. In the end, the bet of this block, in my opinion, is that the combination of _basic and universal blocks_ and _highly specialized blocks_ can be MAGICAL (⊃｡•́‿•̀｡)⊃━☆ﾟ.*･｡ﾟ

Notably with reusable blocks which can be shared as blueprints across pastebins the web over, it is _extra_ important that Gutenberg at its core ships with both the universal blocks, and the dedicated layout blocks.

aduth commented:
I don't disagree that this needs to be simple from a user's perspective. From the technical end, having distinct implementations of blocks for layout variations which can otherwise be represented by existing block configurations is going to quickly become a huge maintenance burden and back-compat concern. It's also avoiding opportunities to solve these needs at a framework level (pre-configured blocks, feature alignment between Columns-like blocks, non-stylistic block variations).

SNaushadS commented:
Loved this enhancement, just waiting for this enhancement in the plugin. @jasmussen has articulated the exact frustration that i was having when trying to get an image and a paragraph.

aduth commented:
Thanks for the context, Jorge. I think we are generally in agreement as far as functionality which could be considered as enhancements to columns, or at least common across columns-like blocks. 

Brain-dump:

It was a worry of mine that rather than considering what might be required to implement these as part of columns, we skip straight toward the most direct solution. That was largely an ignorance of the history on my part, given previous work done in #7414. In retrospect, it's been good to have highlighted some of these needed functionality: resizable columns, pass-through nested blocks, column background, column vertical alignment. Part of my initial reaction was in wondering whether we might anyways want some built-in support for creating a specialized block type which is really just a pre-configured block of another type.

The other concern I have is in making sure that in implementing these as separate blocks, we don't neglect to consider consistency between the blocks. As you noted, this is something we can (and do) achieve through components. In talking with @mtias, there are also cases where we need some combination of rendering implementation of a component, as well as attributes definition. We can see this already in existing components, such as colors, alignment, URL, media selection. We already have some issues with this, at least so far as consistency in the server-side implementation of the block (https://github.com/WordPress/gutenberg/issues/2751#issuecomment-421107960), and even client-side (inconsistent URL input #8000 #8842). It's also a common criticism to blocks like the Button block, which one might consider to be one of the simplest cases, but requires hundreds of lines of code to implement. I'm interested to see what we might be able to do here in exploring whether it might make sense to introduce some concept of a limited block type, which defines some attributes that it merges into its parent context (`core/colored-text`). The idea here is that we can take advantage of portability of components (some already existing e.g. `withColors`), as well as bind attribute handling, ultimately allowing implementations of blocks using these shared functionalities to be simpler and consistent.

Taking this block for example, essentially all of the attributes which are defined specific to the Half Media block seem as though they are really the sum of the composite parts: alignment, color, media selection. Expressing as a composition this way seems like it's probably a better approach than what I was originally recommending with building all of this functionality into the columns block. All we are missing are the tools to express the composition, such that we can avoid hand-coding this each time we implement a new block...

...which, of course, is not something I think we need to address here. My only other concern then would be to make sure that we can have graceful transition processes for if/when these come to be: whether that is ensuring that we have the necessary support in the deprecated process, or we plan to create new blocks and can automatically transition these blocks when those are created.

How best then can we track these needs to be able to move forward with this as is? Is that to be kept as represented by #6993 ?

jorgefilipecosta commented:
Hi @aduth thank you for sharing your thoughts in detail.
Regarding the sharing of logic between blocks, we have the supports: align which automatically adds attributes, UI, and save logic to block that need it. This approach has the limitation of not being flexible and does not allow changing the UI without adding more support flags, also makes impossible for the server to be aware of this attributes.
We have the colors approach where logic is abstracted in withColors HOC and then we provide UI components that allow the consumer to specify colors as they want. And probably we have other approaches.
We will need to check the approaches we have to decide on the one that best fits for this cases and if necessary refactor existing code using approaches to use the chosen approach. This is a complex problem and suggestions for how to implement reusability between blocks are welcome.

Regarding this PR I think there are some things we can make to make this move forward. We need to check if the markup being generated and attributes we have are optimal for this use case (the more eyes on the markup the better). I will probably also not expose MediaContainer component and make it part of the block for now. So there will be no dependencies on the component and if we find a better interface nobody will be affected with the deprecation of the component.
Provided we don't change the markup and the attributes we can apply any change to this block in the future to use our reusable components without requiring deprecations and impacting the users. The attributes may even be inserted by an automatic system in the future. Advancing with the PR right now would allow us to offer this block to the users sooner and improve the UX of the block (if we find it is needed) in parallel to the works being made on the reusability approach.

jasmussen commented:
By the way should we call this block "Media & Text"?

jorgefilipecosta commented:
All the blockers of this PR have been merged :) This PR was rebased and updated to use components that meanwhile have been created and simplified the code.
It is pending a code review before the merge.

jasmussen commented:
![screenshot 2018-10-09 at 13 58 40](https://user-images.githubusercontent.com/1204802/46667997-87af1c80-cbcb-11e8-8df3-b933d6e6acae.png)

Digging the name.

This seems good to me! If someone else gives a code thumbs up, then 👍 👍 from me.

mtias commented:
I'm punting this from 4.0 because it's already a huge release and want to minimize the surface area a bit.

mtias commented:
Thanks, this seems better.

Right aligned looks a bit awkward, though, we should have a similar padding there:
![image](https://user-images.githubusercontent.com/548849/46947616-a86bec00-d07b-11e8-8c0e-640a0e15fa5c.png)

mtias commented:
The `margin-bottom: -10px` also causes a visual glitch of a couple pixels in my testing (Chrome):

![image](https://user-images.githubusercontent.com/548849/46947653-d05b4f80-d07b-11e8-9bc2-b4e90e00b06b.png)

afercia commented:
Looking at this for the first time, there are a few a11y issues to address I've noticed so far.

1
Initial focus. when inserting this block, initial focus goes to the _second_ RichText field:

![screen shot 2018-10-15 at 19 32 58](https://user-images.githubusercontent.com/1682452/46968035-8772bd80-d0b2-11e8-8199-92dee1c9ab07.png)

Tab again and focus goes to the sibling Add block button:

![screen shot 2018-10-15 at 19 33 04](https://user-images.githubusercontent.com/1682452/46968066-96f20680-d0b2-11e8-8842-fe008f98023f.png)

Tab again and focus goes outside of the block to the following one:

![screen shot 2018-10-15 at 19 33 06](https://user-images.githubusercontent.com/1682452/46968092-a40ef580-d0b2-11e8-90ce-3bfaab3b37b4.png)

This makes all the UI that is _before_ the second RichText hardly discoverable for screen reader users and hardly operable for sighted keyboard users. Ideally, I'd tend to think this is one case where setting initial focus on the block outer wrapper would be the best option, also considering that in an already existing block the order of media and text could be reversed.

2
The two editable fields _wrapper_ are both paragraph blocks and announced as such via their aria label `aria-label="Block: Paragraph"` (note: I'm referring to the _wrapper_ not the RichText). Not sure this is ideal because that's not what they are. The first block is sort of a big "title", while the second is normal text but not exactly a paragraph. The first one especially, doesn't give any info to non sighted users that it's a "big" text meant to be used as sort of title. Wondering how the blocks aria-label could be changed in this specific case or maybe just make the bit text a heading?

jorgefilipecosta commented:
Hi, @mtias thank you for your tests both issues mentioned (right padding and visual glitch on the image were addressed).

mtias commented:
Can we remove the second block of text or is that dependant on addressing the trailing inserter issues?

jorgefilipecosta commented:
Hi @afercia, thank you for testing this PR from an a11y stand point. At first sight, it seems some of the problems are caused by existing issues/bugs and/or require engine feature additions so they can be addressed (e.g: aria-labels dependent on the nesting context). I will dive into your feedback to see if we can address in this PR the issues or if they are general ones and we should work on a general solution.

Hi @mtias,
> Can we remove the second block of text or is that dependant on addressing the trailing inserter issues?

We are not inserting the second block of text we just insert the first one. The second block of text is caused by a bug (I think related with the trailing inserter issues) and should not exist at all. This gist https://gist.github.com/jorgefilipecosta/edafb2422ef41020d75619adf31d725e contains a very simple block that shows this bug in action.

jorgefilipecosta commented:
Hi @afercia,

> Ideally, I'd tend to think this is one case where setting initial focus on the block outer wrapper would be the best option, also considering that in an already existing block the order of media and text could be reversed.

I totally agree with this. Currently, we have a limitation where if a block contains a template we never keep the parent selected and we always select one of the children. So this problem is a general one and for any block with an edition area followed by a nested area the edition area would be hard to discover. I created a PR to overcome this general limitation. It allows blocks with templates to be kept selected when inserted.

> The two editable fields wrapper are both paragraph blocks and announced as such via their aria label aria-label="Block: Paragraph" (note: I'm referring to the wrapper not the RichText). Not sure this is ideal because that's not what they are.

Currently, we can not change the behavior of a block depending on the place where it is nested.
For example for an image block nested in an author bio block containing the author picture, we can not disable the alignment controls, even if any that context alignment is not possible and the controls don't make sense. From the a11y point of view for that use case, it would also be helpful to have  aria-label="Block: Image (Author Photo)" instead of aria-label="Block: Image".
We have an issue opened that should allow us to change blocks depending on some variables like the place where they are nested. If we provide that mechanism it should be possible to change aria-labels of the blocks from the outside of the block (e.g.from the parent), so describing what the blocks are would be trivial. 
I added the accessibility label to the existing issue as the issue looks very important from the a11y point of view https://github.com/WordPress/gutenberg/issues/7931.

youknowriad commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -0,0 +1,157 @@
+				// video contain the media type of 'file' in the object returned from the rest api.
+				mediaType = 'video';
+			}
+		} else { // for media selections originated from existing files in the media library.
```
I guess I have the same remark. Do we have an issue to follow up with this? Unify the media object between the MediaUpload and the rest API?

youknowriad commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -0,0 +1,157 @@
+		const { attributes, backgroundColor, setAttributes, setBackgroundColor } = this.props;
+		const { mediaPosition, mediaWidth } = attributes;
+		const temporaryMediaWidth = this.state.mediaWidth;
+		const className = classnames( 'wp-block-media-text', {
```
I think the best practice is to use the "className" prop instead of hardcoding this 'wp-block-media-text'

youknowriad commented on packages/block-library/src/media-text/editor.scss:
```diff
@@ -0,0 +1,61 @@
+
+
```
Should we remove these empty lines?

youknowriad commented on packages/block-library/src/media-text/index.js:
```diff
@@ -0,0 +1,126 @@
+			mediaWidth,
+		} = attributes;
+		const mediaTypeRenders = {
+			image: () => {
```
Minor: could be just `image: () => <img src={ mediaUrl } alt={ mediaAlt } />`

youknowriad reviewed (approved):
Some small remarks, but overall, looking good to me.

I'd suggest doing a quick test in IE11 and ship.

jorgefilipecosta committed:
Revisions

paaljoachim commented:
I am trying to recreate these sections using the block: http://edwien.no/publikasjoner/
Result in the backend:
<img width="809" alt="screen shot 2018-10-22 at 17 54 17" src="https://user-images.githubusercontent.com/5323259/47303201-8f3ae080-d623-11e8-8064-05c09cc82780.png">

Image has space above and below. Heading aligns somewhat to the top of the image. Text links aligns somewhat to the bottom of the image.

In the frontend it looks like this:
<img width="1118" alt="screen shot 2018-10-22 at 17 55 06" src="https://user-images.githubusercontent.com/5323259/47303217-a24db080-d623-11e8-809c-5d2cf782371b.png">

The frontend will show the image nudged up top, left and it is going below the edge of the box.
Heading and text shows up in the center of the image.

lkraav commented:
Testing it today, I wasn't sure why I wasn't able to add a Quote block in the Text container.

webmandesign commented:
Hi guys,

I certainly don't want to sound disrespectful, that's not my intention here. I just have feeling this block might be kind of redundant.

You see, if **Columns** block allowed setting up a _all/each column background and text color_, and _vertical alignment_ (which is definitely possible as columns are using styled using flexbox) I thing we could replicate **Media & Text** block's functionality that way and yet allow more flexibility still (as from what I know I can use any block inside column).

However, I understand having a dedicated **Media & Text** block simplifies the setup of such layout section, as my approach would require some much needed improvements to **Columns** block setup.

That's just my thoughts&hellip;

_Additional suggestion:_ If **Media & Text** block is here to stay, I suggest adding an option to set up also a text color, not just background color.

aduth commented on packages/block-library/src/media-text/media-container.js:
```diff
@@ -0,0 +1,125 @@
+			}
+			return (
+				<ResizableBox
+					className="editor-media-container__resizer"
```
This doesn't adhere to class naming conventions.

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#naming
